### DISCUSSION TITLE
Minimal reproduction instructions/templates

### DISCUSSION THREAD

**Discussion by HonkingGoose** [renovatebot/renovate#28614] [Suggest an Idea]
### Tell us more.

For context read this first:

- https://github.com/renovatebot/renovate/pull/28606#issuecomment-2074257052

We have a [`renovate-reproductions/base`](https://github.com/renovate-reproductions/base/) repository. Do we want people to use this repository as a base to start _their_ reproductions?

I guess not, because that repository has a lot of files. For example, sometimes we only need a `renovate.json` file and a simple `package.json` file to get a working reproduction.

**Comment by viceice** [comment]
no, it's simply another test repo.

**Comment by karlhorky** [comment]
Thanks for opening this discussion!

> > Do we want people to use this repository as a base to start _their_ reproductions?
>
> I guess not, because that repository has a lot of files. 
>
> no, it's simply another test repo.

Ok, I understand that it is perceived as potentially situationally overkill for Renovate maintainers.

Eg. listening to this seems to make sense to me at first:

> For example, sometimes we only need a `renovate.json` file and a simple `package.json` file to get a working reproduction.

One thing that this leaves out is the experience of community members, often those who don't have visibility into:

1. What is the "current version" of best practices for creating a reproduction
2. What is required (or not required) for reproductions of different complexity

So codifying these details into 1 (or 2 or 3, if there are different levels of complexity of reproduction) maintained reproduction templates - even if the simpler level(s) of complexity are only 2 files in a repo - would in my experience in open source communities lead to:

A) higher community engagement, especially from those who are less involved right now
B) reduced back and forth between Renovate maintainers and the community, when reproductions are either not simple enough or do not check certain boxes

Even having contributed a few things to the Renovate codebases over the years, I would have to put in some effort, thinking and tie into "how to create a nice, simple reproduction", which will be a chilling effect to a lot of potential contributions.

## Prior Art

- Next.js CodeSandbox reproduction template https://codesandbox.io/p/sandbox/github/vercel/next.js/tree/canary/examples/reproduction-template
- Testing Library Reproduction Template https://github.com/testing-library/dom-testing-library-template
- Ent Template Repository For Bug Reporting https://github.com/ent/bug
- Reproduction Template of ESLint Stylistic https://github.com/eslint-community/eslint-stylistic-repro-template
- GitHub template for creating a Rspack minimal reproducible example https://github.com/web-infra-dev/rspack-repro
- Renode issue reproduction template https://github.com/renode/renode-issue-reproduction-template
- Vue Test Utils Reproduction Template [issue creation template](https://github.com/vuejs/test-utils/blob/1cdd7ae5a973e8a2de8eef8c22159bc0cc97f911/.github/ISSUE_TEMPLATE/bug_report.md?plain=1#L16-L19) [repro template on StackBlitz](https://stackblitz.com/github/vuejs/create-vue-templates/tree/main/typescript-vitest?file=src%2Fcomponents%2F__tests__%2FHelloWorld.spec.ts)
- shadcn-svelte Reproduction Template https://github.com/huntabyte/shadcn-repro-template
- Ant Design reproduction template [repro template on CodeSandbox](https://codesandbox.io/s/antd-reproduction-template-forked-jyh2k9) [Link to docs](https://ant-design-3x.gitee.io/docs/react/getting-started#1.-Create-one-codesandbox:~:text=Visit%20http%3A//u.ant.design/codesandbox%2Drepro%20to%20create%20a%20codesandbox.%20Don%27t%20forget%20to%20press%20the%20save%20button.)
- Storybook reproduction generator in the CLI https://storybook.js.org/docs/contribute/how-to-reproduce#initial-setup
- Porsche Design reproduction generator with StackBlitz https://twitter.com/stackblitz/status/1648341661762633729
- Apollo GraphQL template for creating minimal reproductions https://github.com/apollographql/next-apollo-example

### Automation of Reproduction Creation

- Bun automatic reproduction generator idea https://twitter.com/jarredsumner/status/1781214396263661985

**Reply by HonkingGoose** [reply]
We do have a guide on minimal reproductions. Which we link to in the auto bot comment asking for a reproduction: [minimal reproductions guide](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md).

**Reply by karlhorky** [reply]
That's good documentation, and good background (although long and detailed - hard to find the information of "how can I create a repro in 30 seconds")

Regardless of how easy it is to read, I would argue that such documentation does not replace a one-click reproduction creator (it could supplement it though).

**Reply by HonkingGoose** [reply]
Another prior art thing: [Material for MkDocs, creating a reproduction](https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/), that we may get inspired by.

**Reply by karlhorky** [reply]
Nice, I expanded on my list of reproduction templates and generators above as well

(also added a new section for "Automation of Reproduction Creation", which is next level automation that the TypeScript team has implemented and other teams are experimenting with)

**Reply by HonkingGoose** [reply]
@karlhorky Is there a repository that lists all these reproduction examples? It would be very nice to have a single, nicely formatted readme, in a "reproductions best practices" repository. A benefit of putting it in a repository is that you can get drive by contributions from users, when the spot a nice new example.

**Reply by karlhorky** [reply]
Good idea - I created a repo here:

- https://github.com/karlhorky/awesome-open-source-automation

Just need to wait 30 days to submit to https://github.com/sindresorhus/awesome

**Reply by HonkingGoose** [reply]
> That's good documentation, and good background (although long and detailed - hard to find the information of "how can I create a repro in 30 seconds")

You're right, we should get to the point quicker. I'll rewrite this soon! :smile: 

- #28730

> Regardless of how easy it is to read, I would argue that such documentation does not replace a one-click reproduction creator (it could supplement it though).

It would be _very nice_ to have a "one click" reproduction system for Renovate. I fear that creating such a thing would be _very hard_, as there are many configurations and platforms that end-users can have.

**Reply by karlhorky** [reply]
> It would be _very nice_ to have a "one click" reproduction system for Renovate. I fear that creating such a thing would be _very hard_, as there are many configurations and platforms that end-users can have.

Just to copy [my issue comment](https://github.com/renovatebot/renovate/issues/28730#issuecomment-2134940549) over to the discussion too:

> Also, for some first baby steps in the direction of [a minimal reproduction template](https://github.com/renovatebot/renovate/discussions/28614#discussioncomment-9211058), I would add an example config link to step&nbsp;3:
> 
> ```diff
> -1. Reduce your Renovate config to a minimum
> +1. Reduce your Renovate config to a minimum (eg. [simple base config starter](https://example.com))
> ```
> 
> Then, if the idea of a minimum reproduction template is amenable to the Renovate team, this could be expanded to multiple GitHub repo templates - one for each for each package manager (eg. including an empty `package.json` for Node.js projects), to get closer to [the "one-click starter" idea](https://github.com/renovatebot/renovate/discussions/28614#discussioncomment-9211953) that I mentioned.

So this may be some first possible steps to create an MVP of a "one-click" reproduction starter

**Reply by HonkingGoose** [reply]
## About reproduction templates for package managers

Do we want to create templates for each package manager, or even the most common managers? Like:

- `npm`/`yarn`/`pnpm`
- `pip`
- `docker` and `docker-compose`
- `github-actions`

Creating the templates is a one-off effort, but then we need to update them when the conventions or ecosystem changes.

Also, with more templates we increase the odds that things get out of date/stale. I don't know if we can monorepo the templates? A bit like this:

- Basic template
  - Folder for `npm`/`yarn`/`pnpm`
  - Folder for `pip`
  - Folder for `docker` and `docker-compose`
  - Folder for `github-actions`

Also having more options might make it harder to find the correct template, or to get overwhelmed with choice.

I'll let a maintainer weigh these points, and decide on more in-depth templates.

## Start with a basic template first?

We could start with a basic template repository, and make the basic setup quicker though.

## Basic template repo

I can create a "template repository" that:

- [ ] Has a `readme.md` file with:
  - [ ] The correct headings, formatting and styling.
  - [ ] A placeholder for a link to the upstream Renovate Issue or Discussion.
- [ ] A minimal `renovate.json` config file which:
  - [ ] Extends `config:recommended`.
  - [ ] Has the schema info (for VSCode auto-complete).
- [ ] Is marked as a template repository.

### Steps for _user_ after forking the template

1. Add the files for their specific problem (say a `package.json` or `package-lock.json` for npm type problems)
3. Update the `renovate.json` file with the correct config to trigger the bug/missing behavior
4. Update the link so it points to the Issue/Discussion
5. Explain Current and Expected behavior

### Where to put the template repo?

After I have created the template repository, we'll have to place it in a location that we can maintain. I'll let the maintainer decide where to put it:

1. In the main Renovate organization.
2. In the Renovate reproductions organization.

### Marking a repository as template

Quote from the [GitHub Docs, creating a template repository](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository):

> Who can use this feature?
> Anyone with admin permissions to a repository can make the repository a template.

## Let me know what you think

Tell me if you like the idea, or not. 😄

**Reply by karlhorky** [reply]
The idea of trying a first version with the most common package managers would make sense to me, to keep initial investment low, track how users interact with these templates, track the maintenance effort of the templates over time.

> we need to update them when the conventions or ecosystem changes

To this point, I can speak a bit:

1. The maintenance effort of `package.json` over time in a template is not huge. I think there have been almost zero actual necessary changes to `package.json` with the Next.js Reproduction Template, which I've been working on for a few years.
2. Package managers occasionally change lockfiles with new versions, which would potentially need an update - if lockfiles were to be a part of a starter template (could also be argued that lockfiles shouldn't be a part of the repro template, because users need to generate a lockfile anyway when adding dependencies)

> I don't know if we can monorepo the templates?

I think having each starter template as a separate repo would be good, to enable easy usage of the GitHub repo-level features "Use template" or "Fork". So something like:

```
- https://github.com/renovate-reproductions/reproduction-npm
- https://github.com/renovate-reproductions/reproduction-yarn
- https://github.com/renovate-reproductions/reproduction-pnpm
- https://github.com/renovate-reproductions/reproduction-pip
- https://github.com/renovate-reproductions/reproduction-docker
- https://github.com/renovate-reproductions/reproduction-docker-compose
- https://github.com/renovate-reproductions/reproduction-github-actions
```

(the reason I included the `reproduction-` prefix is to make the repo name less confusing after forking / using as a template - could also be `renovate-repro-` prefix to be more explicit)

**Comment by rarkins** [comment]
I think it would be great to have some "perfect" examples of reproductions to refer to. They're meant to be minimal, so the example is almost self-documenting - everything is there for a reason

**Reply by HonkingGoose** [reply]
I'm adding a perfect reproduction example with this PR:

- #29316

**Comment by rarkins** [comment]
I hope we can have:

1. A minimal template repository for users to import (not fork, as you can only fork once per repo), e.g. includes README with headings, empty `renovate.json`
2. 1-2 examples of "perfect" reproductions from the past

Hopefully once done the above have very little maintenance. It doesn't matter much if the examples are for bugs which were solved years ago.

**Reply by HonkingGoose** [reply]
For item 1, see this issue:

- #29319

For item 2, I'm already adding _one_ example with this PR:

- #29316
